physical/gcs/gcs.go

```diff
@@ -214,3 +257,271 @@ func (g *GCSBackend) List(prefix string) ([]string, error) {
+	obj := bucket.Object(l.key)
+	attrs, err := obj.Attrs(context.Background())
+
+	canWrite := false
+	conditions := storage.Conditions{DoesNotExist: true}
+	// if the object doesn't exist set the expiration time to the current time + ttl
+	expires := strconv.FormatInt(time.Now().Add(l.ttl).UnixNano(), 10)
+
+	// if the object exists we check its identity and expiration time to determine if we can update it
+	if attrs != nil {
+		// We can only write if:
+		// A. The object doesn't exist
+		// or
+		// B. identity is equal to our identity
+		// or
+		// C. The ttl on the item is <= to the current time
+		if identity, ok := attrs.Metadata["identity"]; ok && identity == l.identity {
+			canWrite = true
+		}
+
+		if ts, ok := attrs.Metadata["expires"]; ok {
+			exp, err := strconv.ParseInt(ts, 10, 64)
+			if err != nil {
+				return errors.New("error parsing unix timestamp")
+			}
+
+			if time.Unix(0, exp).UnixNano() <= time.Now().UnixNano() {
+				canWrite = true
+			}
+		}
+
+		conditions.DoesNotExist = false
+		// we must ensure that another process didn't change the object in the period
+		// between retrieving and writing it back
+		// to do that we add a condition to match the generation of the object
+		// https://cloud.google.com/storage/docs/generations-preconditions
+		conditions.GenerationMatch = attrs.Generation
+		conditions.MetagenerationMatch = attrs.Metageneration
+	}
+
+	if !canWrite && attrs != nil {
+		return errors.New(conditionCheckFailedError)
+	}
+
+	rw := obj.If(conditions).NewWriter(context.Background())
+	rw.ObjectAttrs.Metadata = map[string]string{
+		"expires":  expires,
+		"identity": l.identity,
+	}
+
+	if _, err = rw.Write([]byte(l.value)); err != nil {
+		return err
+	}
+
+	if err = rw.Close(); err != nil {
+		// catch the google api errors when conditions are not met
+		if err.Error() == gcsConditionNotMetError || err.Error() == "googleapi: "+gcsConditionNotMetError {
+			err = errors.New(conditionCheckFailedError)
+		}
+	}
+
+	return err
+}
+
+// watch checks whether the lock has changed in the
+// GCS and closes the leader channel if so.
+// The interval is set by `GCSWatchRetryInterval`.
+// If an error occurs during the check, watch will retry
+// the operation for `GCSWatchRetryMax` times and
+// close the leader channel if it can't succeed.
+func (l *GCSLock) watch(lost chan struct{}) {
+	retries := GCSWatchRetryMax
+	ticker := time.NewTicker(l.watchRetryInterval)
+	bucket := l.backend.client.Bucket(l.backend.bucketName)
+	obj := bucket.Object(l.key)
+
+WatchLoop:
+	for {
+		select {
+		case <-ticker.C:
+			attrs, err := obj.Attrs(context.Background())
+
+			if err != nil {
+				retries--
+				if retries == 0 {
+					break WatchLoop
+				}
+				continue
+			}
+
+			if attrs == nil {
+				break WatchLoop
+			}
+
+			if identity, ok := attrs.Metadata["identity"]; ok {
+				if identity != string(l.identity) {
+					break WatchLoop
+				}
+			}
+		}
+	}
+
+	close(lost)
+}
```

Comment:
We should consider supporting [Transactional](https://github.com/hashicorp/vault/blob/5b4abc09ddd82be99e05264a91b06bc18ab1447a/physical/spanner/spanner.go#L295) too, since it'll speed up boot time.